tests/test_buffer.py

```diff
@@ -598,6 +598,52 @@ def test_SimpleArray_broadcast_slice_ghost_md(self):
                 for k in range(4):
                     self.assertEqual(ndarr2[i, j, k], sarr[i - G, j, k])
 
+    def test_SimpleArray_casting(self):
+        # first check the caster works if the argument is exact the same type
+        array_float64 = modmesh.SimpleArrayFloat64((2, 3, 4))
+        self.assertEqual(modmesh.testhelper.TestSimpleArrayHelper.
+                         test_load_arrayfloat64_from_arrayplex(array_float64),
+                         True)
+
+        # init arrayplex
+        arrayplex_int32 = modmesh.SimpleArray((2, 3, 4), dtype="int32")
+        arrayplex_uint64 = modmesh.SimpleArray((2, 3, 4), dtype="uint64")
+        arrayplex_float64 = modmesh.SimpleArray((2, 3, 4), dtype="float64")
+
+        # check the type is the same with different data types
+        self.assertTrue(type(arrayplex_int32) is type(arrayplex_uint64))
+        self.assertTrue(type(arrayplex_uint64) is type(arrayplex_float64))
+
+        # check if arrayplex can cast to simplearray
+        self.assertEqual(modmesh.testhelper.TestSimpleArrayHelper.
+                         test_load_arrayin32_from_arrayplex(arrayplex_int32),
+                         True)
+
+        # int32 and uint64 are different types
+        with self.assertRaisesRegex(TypeError,
+                                    r"incompatible function arguments"):
+            modmesh.testhelper.TestSimpleArrayHelper \
+                .test_load_arrayin32_from_arrayplex(arrayplex_uint64)
+
+        # check if arrayplex can cast to simplearray
+        self.assertEqual(
+            modmesh.testhelper.TestSimpleArrayHelper
+            .test_load_arrayfloat64_from_arrayplex(arrayplex_float64),
+            True)
+
+        # float64 and int32 are differet types
+        with self.assertRaisesRegex(TypeError,
+                                    r"incompatible function arguments"):
+            modmesh.testhelper.TestSimpleArrayHelper \
+                .test_load_arrayfloat64_from_arrayplex(arrayplex_int32)
+
+        # explicitly check the `cast` function of the customized caster works
+        array_int32 = modmesh.SimpleArrayInt32((2, 3, 4))
+        array_int32_2 = modmesh.testhelper \
+            .TestSimpleArrayHelper \
+            .test_cast_to_arrayint32()  # SimpleArray32 from casting
+        self.assertTrue(type(array_int32) is type(array_int32_2))
+
```

Comment:
The following does not seem to work (I am ok with it now):

```python
# Obtain a plex from typed array
array_int32 = modmesh.SimpleArrayInt32((2, 3, 4))
arrayplex_int32 = array_int32.plex

# Obtain the typed array from a plex
arrayplex_int32 = modmesh.SimpleArray((2, 3, 4), dtype='int32')
array_int32 = arrayplex_int32.typed
```

How can we make the above work?

Reply:
yes, we can do it. already added in the PR. will leave explanation comments later.